outline.md

```diff
@@ -149,6 +149,89 @@ Note also that GHC's constraint solver can be called for many different reasons,
+
+For example, suppose we define a nullary type class and a definition which requires it:
+
+```haskell
+{-# LANGUAGE MultiParamTypeClasses #-}
+module My.Nullary where
+
+class MyNullary
+
+requireMyNullary :: MyNullary => ()
+requireMyNullary = ()
+```
+
+If we try to call `requireMyNullary` from a function which does not itself require `MyNullary`, we get a `No instance for MyNullary` error.
+
+```haskell
+{-# OPTIONS_GHC -fplugin My.Plugin
+                -fplugin-opt=My.Plugin:Nullary
+  #-}
+module My.Module where
+
+import My.Nullary (requireMyNullary)
+
+unit :: ()
+unit = requireMyNullary
+```
+
+We can get ghc to accept this code by writing a plugin which discharges the wanted `MyNullary` constraints it encounters:
+
+```haskell
+module My.Plugin (plugin) where
+
+import Class (Class(className, classTyCon))
+import FastString (mkFastString)
+import GHC.TcPluginM.Extra (lookupModule, lookupName)
+import MkCore (mkCoreConApps)
+import Module (mkModuleName)
+import Name (Name, mkTcOcc)
+import Plugins (Plugin(pluginRecompile, tcPlugin), defaultPlugin, purePlugin)
+import TcEvidence (EvTerm(EvExpr))
+import TcPluginM (TcPluginM)
+import TcRnTypes (Ct(CDictCan), TcPlugin(TcPlugin, tcPluginInit, tcPluginSolve, tcPluginStop) , TcPluginResult(TcPluginOk)
+  )
+import TyCon (tyConDataCons_maybe)
+
+plugin :: Plugin
+plugin = defaultPlugin
+  { tcPlugin = \_ -> Just $ TcPlugin
+    { tcPluginInit  = pluginInit
+    , tcPluginSolve = pluginSolve
+    , tcPluginStop  = \_ -> pure ()
+    }
+  , pluginRecompile = purePlugin
+  }
+
+pluginInit :: TcPluginM Name
+pluginInit = do
+  myNullaryModule <- lookupModule (mkModuleName "My.Nullary") (mkFastString "my-package")
+  lookupName myNullaryModule (mkTcOcc "MyNullary")
+
+pluginSolve :: Name -> [Ct] -> [Ct] -> [Ct] -> TcPluginM TcPluginResult
+pluginSolve myNullaryName _ _ = go
+  where
+    go :: [Ct] -> TcPluginM TcPluginResult
+    go [] = do
+      pure $ TcPluginOk [] []
+    go (ct@(CDictCan _ class_ _ _) : _) | className class_ == myNullaryName = do
+      let myNullaryClass = class_
+          dictTyCon = classTyCon myNullaryClass
+          Just [dictDataCon] = tyConDataCons_maybe dictTyCon
+      pure $ TcPluginOk [(EvExpr $ mkCoreConApps dictDataCon [], ct)] []
+    go (_ : cts) = do
+      go cts
+```
+
+In `pluginInit`, we first lookup `MyNullary` in the `My.Nullary` module in order to obtain its internal name. Later on, in `go`, we compare internal names in order to make sure we only discharge our `MyNullary` constraint, and not e.g. some other constraint which also happens to be called "MyNullary". We then discharge the constraint by returning it along with the relevant evidence: a dictionary providing an implementation for all the methods of the type class. Since `MyNullary` has zero methods, we can do this by using `mkCoreConApps` to apply the dictionary's data constructor to zero arguments.
```

Comment:
Perhaps add a reference to the section on typeclass evidence (`#classes`).

Reply:
added